checker/tests/tainting/Issue4170.java

```diff
@@ -0,0 +1,21 @@
+import java.util.ArrayList;
+import org.checkerframework.checker.tainting.qual.Tainted;
+import org.checkerframework.checker.tainting.qual.Untainted;
+
+// @below-java10-jdk-skip-test
+public class Issue4170 {
+  public void method1() {
+    var list = new ArrayList<@Untainted String>();
```

Comment:
Please add tests where the variables declared with `var` are assigned to a variable with a type.  For example, add `ArrayList<@Untainted String> list2 = list` here.

Reply:
`ArrayList<@Untainted String> list2 = list` test is failing with an assignment error. the type of `list` stays `ArrayList<String>` in this line. is there any extra step necessary to persist the type of list from the previous line?

Reply:
I added that line to this test case, and it is not failing for me.  Though when I run with JDK 11, I get a NoSuchMethodException.

Reply:
Could you please share your jdk version where it's passing? I'm trying with `openjdk version "17.0.7" 2023-04-18` and it's failing for this version.

Reply:
I tried again with Eclipse Temurin JDK. It is failing with both JDK 11 and JDK 17 for me. I didn't encounter the NoSuchMethodException with this version of JDK.

Reply:
> I tried again with Eclipse Temurin JDK. It is failing with both JDK 11 and JDK 17 for me. I didn't encounter the NoSuchMethodException with this version of JDK.

FWIW, I see the same result on JDK 11 and 17 on Mac (failed test, no `NoSuchMethodException`).

Reply:
So it seem like `isVariableTreeDeclaredUsingVar` isn't working correctly.  I'm guessing on some JDK's it is returning `false` and on others it is crashing.  @kanaksad you can see if it is returning `false` when it should be return `true`? We may need to come up with anther way to check if a variable is declared with `var`.

Reply:
I have one JDK where it's crashing(_11.0.11 (x86_64) "AdoptOpenJDK"_). But where it doesn't crash(_Eclipse Temurin JDK 11, 17, OpenJDK 17_), the `isVariableTreeDeclaredUsingVar` behavior is as expected, returning `true` where it should return `true`.

Reply:
is it possible that the annotatedtype calculated here is used for type checking only at current node and doesn't update the tree? asking this because [here](https://github.com/typetools/checker-framework/blob/c1edbbb196decfcfc3bdce3e874bff06bfc56852/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java#L1510) the variable type isn't propagated to the super call and type of tree remains the same.

Reply:
@smillst do you have any further tips here?  I think we're not sure what the next step should be

Reply:
I'm still concerned that we are getting crashes on some JDKs, that should be fixed before we merge.  

But, it looks like PropagationTreeAnnotator isn't the right place for this code.  I think it needs to go in `TypeFromMemberVisitor#visitVariable`

Reply:
@smillst I updated the code in TypeFromMemberVisitor#visitVariable and pushed a commit. This works fine with tainting annotations. However, it's still problematic for nullness annotations. 
To retrieve the annotated initializer type, I'm using `TypeFromTree.fromExpression(f, variableTree.getInitializer())`. 
But when `aTypeFactory` is `KeyForAnnotatedTypeFactory`, we're losing annotations from this API. For example, 
```
new ArrayList<@NonNull String>(); // API returns ArrayList<String>
new ArrayList<@Nullable String>(); // API returns ArrayList<String>
```
Also, the API invocation here suffers from the infinite recursion mentioned previously, so the blocking code in `KeyForPropagator#propagateNewClassTree` needs to be there. 

When `KeyForAnnotatedTypeFactory` is in action, I don't see any way to get the annotated type for `NewClassTree` that preserves the annotations. Is there any alternate way you can think of?
The last commit has a few failing tests demonstrating this problem.

Reply:
You should call `f.getAnnotatedType(variableTree.getInitializer())` not `TypeFromTree.fromExpression(f, variableTree.getInitializer())`.   `TypeFromTree.fromExpression(f, variableTree.getInitializer())` does not return the type with all the correct annotations added.

Reply:
@smillst thanks, it solves the problem. there is one new error reported in the existing `all-systems/VarKeyWord.java` test which looks okay to me. can you please confirm?

```
var s = "Hello!";
s = null; // new error reported on this assignment. 
```

Reply:
I just pushed a fix for that testcase.